src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java

```diff
@@ -1067,6 +1067,12 @@ private void createFlattenedDependenciesAll( List<Dependency> projectDependencie
                 {
                     return true;
                 }
+
+                if ("provided".equals(node.getArtifact().getScope())) {
+                    DependencyNode parent = node.getParent();
+                    if(!parent.getArtifact().getGroupId().equals(projectArtifact.getGroupId()) || !parent.getArtifact().getArtifactId().equals(projectArtifact.getArtifactId()))
```

Comment:
nit: space after if

Reply:
Is there a more general case where:

If we are flattening `A` where `A -> B (provided) -> C`
And we have consumer `X` where `X -> A -> B (provided) -> C`

* `B` will not be in the transitive. So it doesn't need to be in the runtime scope. Initial thinking is - we can simply drop `B` and its subtree altogether, just like the `optional` case.
* or, we can keep the information of `provided` scope dependencies in the flattened pom, so that in case a user sees a runtime classpath issue, user can know their environment is missing the `provided` artifacts.

In which case, do we keep only the first level one? or can the code be generic, and simply walk the tree and simply add the `provided` dependency, and return false to drop the children?

If we take that approach, what happens in the case of: `A -> B -> C (provided) -> D`
In this case, C is transitive. C doesn't show up in the compile path, but will we walk to C in this case? If we did, the flattened pom will look like `A -> B, C (provided)`, is that the right behavior?

Reply:
Given `A flattened -> B (provided) -> C`,
It may be better to maintain B but drop its subtree for information purposes. Namely if there's an issue with provided dependencies (e.g. it's not actually provided at runtime), the user will benefit from having additional information regarding the provided artifact. This is particularly useful in the case of differing dependency trees for the flattened POM and the original POM. 

In the case of the code that is currently in the PR (where A is flattened),
`A -> B (provided) -> C` <==> Output: `A -> B (provided)`

I believe this should be the desired output as well, since we maintain our direct dependency on B from the flatten perspective. From the consumer perspective, it also gives a clear guideline for output (all second-level transitive provided dependencies are dropped, while first-level transitive provided dependencies are kept when compiling with the flattened POM). A good example of this would be the proto-javabigquerystorage / grpc-javabigquerystorage libraries.

It's also good for informational purposes, as mentioned with potential classpath issues, in which the consumer would receive more information on where their usage of provided dependencies might be going wrong with their provided dependencies. 

--------

In most cases we could just walk the tree, add the associated provided dependency to our linked list, then return false to not include its subtree. This would work for all cases except when a transitive node is managed to the provided scope (with a compile scope parent) ==> `A -> B (compile) -> C (managed to 'provided')`

In this case, we would need to know whether the node was transitive or not, meaning we would either have to look up from the node itself or look down from its parent.

Reply:
thanks for the investigation and the clarification. I feel this is sound. The only last nits are the code format to be consistent w/ the rest of the file, i.e.,
```
if (....)
{
  ...
}
```

Reply:
All formatting has been fixed. Thank you for the notification!